File: go/two_pointers/125_valid_palindrome.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

func main() {
	// Test cases

	// result: true
	fmt.Println(isPalindrome("A man, a plan, a canal: Panama"))

	// result: false
	fmt.Println(isPalindrome("race a car"))

	// result: true
	fmt.Println(isPalindrome(" "))

	// result: true
	fmt.Println(isPalindrome("aa"))
}

func isPalindrome(s string) bool {
	nonAlphanumeric := regexp.MustCompile("[^a-z0-9]+")
	palindrome := nonAlphanumeric.ReplaceAllString(strings.ToLower(s), "")

	if len(palindrome) <= 1 {
		return true
	}

	left := 0
	right := len(palindrome) - 1

	for right >= left {
		if string(palindrome[left]) != string(palindrome[right]) {
			return false
		}
		left += 1
		right -= 1
	}
	return true
}

File: js/two_pointers/125_valid_palindrome.ts
function isPalindrome(s: string): boolean {
  const palindrome: string = s.toLowerCase().replace(/[^a-z0-9]/gi, '');
  if (palindrome.length <= 1) return true;

  let left = 0;
  let right = palindrome.length - 1;

  while (right >= left) {
    if (palindrome[left] !== palindrome[right]) return false;

    left++;
    right--;
  }

  return true;
}

(() => {
  // Test cases

  // result: true
  console.log(isPalindrome('A man, a plan, a canal: Panama'));
  // result: false
  console.log(isPalindrome('race a car'));
  // result: true
  console.log(isPalindrome(' '));
  // result: true
  console.log(isPalindrome('aa'));
})();
